### PR TITLE
Featuring omrabbitmq

### PR DESCRIPTION
**Pull request by pduveau**
- add routing-key template to allow routing depending on the message
- add capability off populating amqp header
- add delivery mode
- add expiration
- add HA capability with two rabbitmq servers

### CODE REVIEW

**Comment by rgerhards** [comment]
@pduveau can you tell me which dependencies are necessary to build omrabbitmq? I would then install these in the CI containers.

**Comment by rgerhards** [comment]
I think we see quite some CI false positives, possibly induced by too-busy CI system (known issue which I am working on). But anyhow, we need to update CI to actually compile (and possibly test) `omrabbitmq`. It would be great if we could at least build.

For this reason, I do not re-init CI as we need to do that in any case once CI is upgraded.

**Comment by rgerhards** [comment]
I have now enabled omrabbitmq in parts of the CI -- found the necessary dependency.

**Comment by pduveau** [comment]
I forget to move new values for RABBITMQ in "enum rsRetVal_" to the right PR. I'll change this. The first PR was not simple to split in smaller ones.

**Comment by rgerhards** [comment]
> I forget to move new values for RABBITMQ in "enum rsRetVal_" to the right PR. I'll change this. The first PR was not simple to split in smaller ones.

excellent!

If you can open the PRs for the other modules, I can also update CI for them, so that you have a quicker feedback. Just a thought...

**Comment by pduveau** [comment]
Hi @rgerhards, I have lot of problems with lines too long because I use lot of indention. My problem is aI think because indention are defined as equivalent to 8 spaces. Is it possible to reduce this equivalence to 4 spaces.

**Comment by davidelang** [comment]
changing the tab to 4 spaces doesn't really address the problem.

please look at your code,does it make things easier to read if you take some of 
the more indented stuff and change it into a helper function?

David Lang

**Comment by rgerhards** [comment]
> I have lot of problems with lines too long because I use lot of indention. My problem is aI think because indention are defined as equivalent to 8 spaces. Is it possible to reduce this equivalence to 4 spaces.

well, than the lines - in standard settings - would still be too long. As @davidelang said, this is also often the result of using too complex functions, which should be refactored.

**Comment by davidelang** [comment]
please note that this is not just a knee jerk coding style restriction. When we 
set the coding style, I raised this very issue, and the max line length was 
lengthened significantly to allow for more indentation.

I don't buy into the 'every function must fit in  a 80x24 screen and have at 
least 5 lines of documentation at the top' mentality, but functions should be 
readable, and if there is too much going on, it becomes hard to track what's 
happening.

**Comment by pduveau** [comment]
99.9 % of time this rule is the best.
Unfortunately, the function, which raise the limit, is a connection monitoring with nested if/switch/case computing connection state's vars. I tried to define an helper function but it had a lot of args and a part of them pointer to the state's vars. I rolled back as the result was worst for readers.  
Also I have to deal with the RabbitMQ library methods and constants which are 20-40 chars each. This makes you quickly reach the limits. ;-)
I'll try to use other strategy to deal with limitations in a nicer way.
B.R.

**Comment by davidelang** [comment]
If the problem is a series of if-then-else statements, then indent them as if 
they were a case statement.

That makes the formatting match the intent far better than nesting each option 
further to the right.

David Lang

**Comment by pduveau** [comment]
Can you give me more codecov/project check ? I can't understand through the dashboard why this check failed ?

**Comment by rgerhards** [comment]
Looks like this is the result of the test bench flake. I have re-initiated. CI has problems right now if very busy. And these days it is very busy. We are working on those flakes.

**Comment by pduveau** [comment]
@rgerhards, could you have a look to failed checks, please ? And relaunch them if possible. Thanks a lot.

**Comment by rgerhards** [comment]
@pduveau done - one test is waiting on a free builder

**Comment by pduveau** [comment]
@rgerhards, I am sorry to bother you again. Just the last failed check to (re)relaunch.

**Comment by rgerhards** [comment]
I just relalize that this module also does no run through static analyzer.... let's see

File: contrib/omrabbitmq/Readme [hunk 1]
```diff
@@ -0,0 +1 @@
+Force PR to pass
```
Comment (rgerhards) [inline contrib/omrabbitmq/Readme]: I guess this file needs to go away.

side-note:
```
$ git commit --amend
$ git push -f
```
also forces new full CI run.

Comment (pduveau) [reply contrib/omrabbitmq/Readme]: Change done

**Comment by rgerhards** [comment]
ubuntu 16 is unrelated, see https://github.com/rsyslog/rsyslog/pull/3504

**Comment by pduveau** [comment]
Something went wrong with ksi code something became deprecated certainly due to library update. You can that in this first check in "continuous-integration/travis-ci/pr"

**Comment by rgerhards** [comment]
I know, maybe I posted to one of your other PRs. It's sorted out, see https://github.com/rsyslog/rsyslog/pull/3507

I just need to wait for merge.

**Comment by rgerhards** [comment]
doc PR: https://github.com/rsyslog/rsyslog-doc/pull/780

**Comment by rgerhards** [comment]
@pduveau could you pls do a rebase?

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging c4e418b534a0b83d3da95396f3caa98e547236f8 into e25edc8354d576e416f9a004c2759806011ca424 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-17a9b587dd1fb504715544dec03b16e012f55d67)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging 2116a647aeb3f72aa6c71803f7e9917e65b1b35a into caf73d452d696b5da27f6e15bf9a4376a62f3d29 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-83947f34a268b87962c232c98df943fe4c24fc96)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging bfac6bfc4165534d25e9f2d37290c75d113643bc into caf73d452d696b5da27f6e15bf9a4376a62f3d29 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-697340e8b80c35a931c56af25f8322bc499aa804)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: tests/omrabbitmq-injector_via_improg.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
```
Comment (rgerhards) [inline tests/omrabbitmq-injector_via_improg.sh]: just to keep test complexity low: can't we use the regular methods (via tcpflood or injectmsg) to inject messages? Having improg run here considerably increases the tested set and so may lead to failures from other components. Just a thought...

Comment (pduveau) [reply tests/omrabbitmq-injector_via_improg.sh]: This is a very good notice. Until, now the problem is not there. I will simplify the inject as soon as a get a stable way of testing output. Again docker-ubuntu18-san is a "punishment" for me. I have not yet understand what's wrong in my tests infra. Do I have a way to detect this particular configuration to skip some tests on if I am not able to find why I get problems there.

Comment (pduveau) [reply tests/omrabbitmq-injector_via_improg.sh]: I used improg just because I knew it.

Comment (rgerhards) [reply tests/omrabbitmq-injector_via_improg.sh]: > Do I have a way to detect this particular configuration to skip some tests on if I am not able to find why I get problems there.

You can skip a platform, but you cannot skip a specific test run - after all, that means the test fails on that platform (here: ubuntu18). That it works in some configurations doesn't help. So we need to find out what the issue really is. Let's do that on the main discussion thread, not this review comment.

**Comment by rgerhards** [comment]
@pduveau in the test log I see
```
./miniamqpsrvr -p 56239 -f rstb_731875_3d5255f8.amqp.log -e rstb_731875_3d5255f8.amqp.err -d &
00.003 Server AMQP 2 thread could not be created (error:0) !
00.003 Server AMQP 1 on port 56239 started
```
Is that OK? Or is it where the problem begins? I see similar messages also later in the log.

I also see this here:
```
rsyslogd: omrabbitmq module 0/1: server localhost port 56239. [v8.1904.0.master try https://www.rsyslog.com/e/2449 ]
rsyslogd: omrabbitmq module connection failed 3 times on each server. [v8.1904.0.master try https://www.rsyslog.com/e/2447 
```
so it kind of looks like the `miniamqpsrvr` does not work? I would suggest we first add a self-test for that tool and/or make it's output more verbose. Maybe I just don't understand what happens here ;-)

Are there some standard tools in rabbitmq for a minimal client and server (like kafkacat in kafka)?

**Comment by rgerhards** [comment]
umm... something is strange. It looks like ubuntu18-codecov does not even run the omrabbitmq tests?

**Comment by rgerhards** [comment]
@pduveau I found one issue that causes omrabbitmq not to be checked under "make distcheck". Let's see if we get at least a second failing worker. Again, feel free to squash my commit. We need to check this in the other contributions as well.

**Comment by rgerhards** [comment]
It looks like `omrabbitmq` is not yet enabled on most CI workers. I'll add more, but have a look at the CI issues (I'll try to restart flakes during the evening, but no 100% promise).

**Comment by rgerhards** [comment]
@pduveau Something ist fishy here - CentOS 7 vm also does Not Run the Tests even though it compiles the module.

**Comment by pduveau** [comment]
"00.003 Server AMQP 2 thread could not be created (error:0) !" is a noisy log (a miss condition test) because there is only server configured (-q is not provided on command line) so the errno = 0 is normal : no error. I have to patch this.
I don't understand why this is not running in Centos 7. I am using Rhel 7 to make my tests. And It is 100% operationnal on. The difference is. My VM is dedicated to me and I am root on. I will try to get environments closed to your. Could you send me the major configuration datas (the limits of user used to test, ...)

**Comment by rgerhards** [comment]
> I don't understand why this is not running in Centos 7. I am using Rhel 7 to make my tests. And It is 100% operationnal on. The difference is. My VM is dedicated to me and I am working as root on. I will try to get environments closed to yours. Could you send me the major configuration datas (the limits of user used to test, ...) for the ubuntu18-*

did you know you can easily run in exactly the same env? All you need is a docker install. These are all containers and so you can run them on you local machine (well, of course the VMs not).  The container definitions for Ubuntu 18 are here: https://github.com/rsyslog/rsyslog-docker/tree/master/dev_env/ubuntu/base/18.04 - this is the sole system config.

I'll try to update the CentOS 7 container today. Unfortunately I am very busy with some other things, so I am not sure how far I come today :-(

**Comment by rgerhards** [comment]
`make distcheck` failed due to missing `EXTRA_DIST` targets; have corrected this with another commit.

**Comment by rgerhards** [comment]
@pduveau CI seems not yet to catch some compile issues with the testbench tools (I guess "make check" overrides the error):
```
../../../tests/miniamqpsrvr.c: In function 'amqp_srvr':
../../../tests/miniamqpsrvr.c:254:11: warning: 'frame.ch' may be used uninitialized in this function [-Wmaybe-uninitialized]
  buf[1] = (char) (channel >> 8);
           ^~~~~~~~~~~~~~~~~~~~~
../../../tests/miniamqpsrvr.c:309:20: note: 'frame.ch' was declared here
  amqp_frame_type_t frame;
                    ^~~~~
../../../tests/miniamqpsrvr.c:510:14: warning: 'body_size' may be used uninitialized in this function [-Wmaybe-uninitialized]
    body_size -= frame.datalen;
              ^~
../../../tests/miniamqpsrvr.c:533:6: warning: 'nSent' may be used uninitialized in this function [-Wmaybe-uninitialized]
   if (nSent < 0) {
```
maybe this is the root cause?

**Comment by rgerhards** [comment]
@pduveau I made a mistake in Ubuntu 18 container build - pushed to wrong label, so actually this was not really completely done. Fixing now...

**Comment by rgerhards** [comment]
@pduveau will somebody take care of this PR next week? Or have we lost?

**Comment by pduveau** [comment]
Hi @rgerhards, I got it. The problem was around thread creation with sometime a successfull thread creation but no thread routine code executed for seconds. Very strange. So now the miniamqpsrvr uses forks and pipes instead of thread and shared memory. This module with tests are complete now. The remaining failed check seems to be systematic. I think you should check it.

**Comment by rgerhards** [comment]
> The problem was around thread creation with sometime a successfull thread creation but no thread routine code executed for seconds. Very strange.

It's actually how pthreads is even supposed to work. I had to learn it the hard way, too. You really need to sync everything where you need a specific order of events.

**Comment by rgerhards** [comment]
I pushed a follow-up PR to enable omrabbitmq in LGTM. Not sure if there was a reason for not having it. I'll monitor the outcome.

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging 483a5de7f6673c9cee3fc5999b2a9b26c0b4b31c into 8c2d76ed4c49ce43f74caac68c4c34054b85610a - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-d1917a82c2fe66389b46eb50c665979fe52f8dbf)

**new alerts:**

* 1 for Empty branch of conditional

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
very happy to finally merge this! Great collaboration! A very big merci beaucoup to @pduveau!

**Pull request merged**